crates/gpui/src/elements/animation.rs

```diff
@@ -165,6 +255,139 @@ impl<E: IntoElement + 'static> Element for AnimationElement<E> {
     }
 }
 
+/// A GPUI element that applies an animation to another element when transitioning between states
+pub struct TransitionElement<E> {
```

Comment:
Similarly, this doesn't _feel_ like a situation that should need a whole new element, that mostly (but not quite) reimplements the existing animation element. I can feel a more general structure under these two but I don't have a precise description yet...